File: plugins/mirage/skills/mirage-vfs-authoring/assets/adapter.py
import asyncio

from mirage import (Accessor, FileStat, FileType, GenericVFS, PathSpec,
                    ReadFixture, ReadOps, VFSAdapter, Workspace,
                    check_read_contract)


class ResourceClient(Accessor):

    def __init__(self) -> None:
        self.files = {"hello.txt": b"Hello from my resource!\n"}


async def read_bytes(client: ResourceClient,
                     path: PathSpec,
                     index=None) -> bytes:
    key = path.vfs_path.strip("/")
    if not key:
        raise IsADirectoryError(path.virtual)
    if key not in client.files:
        raise FileNotFoundError(path.virtual)
    return client.files[key]


async def readdir(client: ResourceClient,
                  path: PathSpec,
                  index=None) -> list[str]:
    if path.vfs_path.strip("/"):
        await read_bytes(client, path, index)
        raise NotADirectoryError(path.virtual)
    return [
        f"{path.virtual.rstrip('/')}/{name}" for name in sorted(client.files)
    ]


async def stat(client: ResourceClient, path: PathSpec, index=None) -> FileStat:
    name = path.virtual.rstrip("/").rsplit("/", 1)[-1] or "/"
    if not path.vfs_path.strip("/"):
        return FileStat(name=name, type=FileType.DIRECTORY)
    data = await read_bytes(client, path, index)
    return FileStat(name=name, type=FileType.FILE, size=len(data))


ADAPTER = VFSAdapter(
    read=ReadOps(readdir=readdir, read_bytes=read_bytes, stat=stat))


async def main() -> None:
    client = ResourceClient()
    fixture = ReadFixture(
        file=PathSpec(virtual="/resource/hello.txt",
                      directory="/resource",
                      vfs_path="hello.txt"),
        directory=PathSpec(virtual="/resource", directory="/", vfs_path=""),
        missing=PathSpec(virtual="/resource/missing",
                         directory="/resource",
                         vfs_path="missing"),
        content=client.files["hello.txt"],
    )
    await check_read_contract(ADAPTER, client, fixture)
    ws = Workspace({
        "/resource":
        GenericVFS(name="resource", accessor=client, io=ADAPTER)
    })
    try:
        result = await ws.shell("cat /resource/hello.txt")
        assert result.exit_code == 0
        assert await result.stdout_str() == fixture.content.decode()
    finally:
        await ws.close()


if __name__ == "__main__":
    asyncio.run(main())

File: plugins/mirage/skills/mirage-vfs-authoring/assets/adapter.ts
import {
  Accessor, FileStat, FileType, GenericVFS, PathSpec, VFSAdapter, Workspace,
  checkReadContract, eisdir, enoent, enotdir,
} from '@struktoai/mirage-node'

const enc = new TextEncoder()

class ResourceClient extends Accessor {
  readonly files = new Map([['hello.txt', enc.encode('Hello from my resource!\n')]])
}

async function readBytes(client: ResourceClient, path: PathSpec): Promise<Uint8Array> {
  const key = path.vfsPath.replace(/^\/+|\/+$/g, '')
  if (key === '') throw eisdir(path)
  const data = client.files.get(key)
  if (data === undefined) throw enoent(path)
  return data
}

async function readdir(client: ResourceClient, path: PathSpec): Promise<string[]> {
  if (path.vfsPath.replace(/^\/+|\/+$/g, '') !== '') {
    await readBytes(client, path)
    throw enotdir(path)
  }
  return [...client.files.keys()].sort().map(name => `${path.virtual.replace(/\/+$/, '')}/${name}`)
}

async function stat(client: ResourceClient, path: PathSpec): Promise<FileStat> {
  const name = path.virtual.replace(/\/+$/, '').split('/').pop() || '/'
  if (path.vfsPath.replace(/^\/+|\/+$/g, '') === '') return new FileStat({ name, type: FileType.DIRECTORY })
  const data = await readBytes(client, path)
  return new FileStat({ name, type: FileType.FILE, size: data.length })
}

export const adapter = new VFSAdapter({ read: { readdir, readBytes, stat } })

async function main(): Promise<void> {
  const client = new ResourceClient()
  await checkReadContract(adapter, client, {
    file: new PathSpec({ virtual: '/resource/hello.txt', directory: '/resource', vfsPath: 'hello.txt' }),
    directory: new PathSpec({ virtual: '/resource', directory: '/', vfsPath: '' }),
    missing: new PathSpec({ virtual: '/resource/missing', directory: '/resource', vfsPath: 'missing' }),
    content: enc.encode('Hello from my resource!\n'),
  })
  const ws = new Workspace({ '/resource': new GenericVFS({ name: 'resource', accessor: client, io: adapter }) })
  try {
    const result = await ws.shell('cat /resource/hello.txt')
    if (result.exitCode !== 0) throw new Error('starter mount failed')
  } finally {
    await ws.close()
  }
}

await main()
